### PR TITLE
Add returnNFT

### PR DESCRIPTION
**Pull request by amarjandu**
Add assertion for ownership
Add body for returnNFT

### CODE REVIEW

File: contracts/Kasu.sol [hunk 1]
```diff
@@ -92,17 +93,41 @@ contract Kasu is KasuStorage {
...
-
+        Listing memory listing = _getListingById(listingId);
+        require(
+            _assertOwnership(
+                listing.tokenAddress,
+                listing.lenderAddress,
+                listing.tokenId
+            ) == true,
+            "Token not returned yet"
+        );
```
Comment (forcey) [inline contracts/Kasu.sol]: I understand that you're trying to check if the NFT has been transferred to the original owner. But ideally we want this to be an atomic transaction:
1. call IERC721.safeTransferFrom() to transfer the NFT (assuming the contract has been approved)
2. transfer collateral to the borrower
3. transfer interest to the lender
4. delete the listing from storage.

Therefore you can make sure there are no loose ends (e.g. what if the borrower returns the NFT, but didn't call returnNFT() and was still charged interest?)

File: contracts/Kasu.sol [hunk 2]
```diff
@@ -92,17 +93,41 @@ contract Kasu is KasuStorage {
...
+        // TODO verify the property houses the collateral amount, what about interest?
+        require(
+            address(this).balance <= listing.collateralRequired,
+            "Insufficient ether in contractAddress"
+        );
+        listing.lenderAddress.transfer(listing.collateralRequired);
```
Comment (forcey) [inline contracts/Kasu.sol]: collateral should go to the borrower on the normal return path. Interest goes to the lender.

File: contracts/Kasu.sol [hunk 3]
```diff
@@ -119,18 +120,34 @@ contract Kasu is KasuStorage, KasuMath {
...
     // [Feature 3] Borrower's dashboard
     // After borrower return NFT, collateral is sent from smart contract to borrower's address
-    function returnNFT(uint256 listingId) public {
-
+    function returnNFT(uint256 listingId) public payable{
+        Listing memory listing = _getListingById(listingId);
+        IERC721(listing.tokenAddress).safeTransferFrom(listing.rental.borrowerAddress,
+                                                       listing.lenderAddress,
+                                                       listing.tokenId);
+        payable(listing.rental.borrowerAddress).call{value: listing.collateralRequired};
```
Comment (forcey) [inline contracts/Kasu.sol]: Looking great!
One remaining issue is that the upfront payment includes interest for the entire rental period; and here the lender is only paid for the elapsed period. So there is a bit of remaining interest money "stuck" in the contract. I think that should be refunded to the borrower along with the collateral.
Feel free to merge this change though, and track this issue in a TODO.

File: src/routes/return.js [hunk 4]
```diff
@@ -1,52 +1,52 @@
...
 import LoginService from '../utils/LoginService';
+import {KasuContract, ERC721_ABI} from '../utils/abiManager';
 
 function ReturnPage() {
     const [rentedNFTs, setRentedNFTs] = useState([]);
-    const fetchRentedNFTs = useCallback(() => {
+    const returnNFT = async function(listing){
+        const loginService = LoginService.getInstance();
+        const tokenContract = new ethers.Contract(listing.tokenAddress, ERC721_ABI(), loginService.signer);
+        await tokenContract.approve(listing.lenderAddress, listing.tokenId)
```
Comment (amarjandu) [inline src/routes/return.js:13]: During local debugging this causes an error like:
```
eth_call
  Contract call:       <UnrecognizedContract>
  From:                0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266
  To:                  0xf953b3a269d80e3eb0f2947630da976b896a8c5b

  Error: Transaction reverted without a reason string
      at <UnrecognizedContract>.<unknown> (0xf953b3a269d80e3eb0f2947630da976b896a8c5b)
      at HardhatNode.runCall (/Users/amar/dev/personal/crypto/nft-rental-marketplace/node_modules/hardhat/src/internal/hardhat-network/provider/node.ts:535:20)
      at EthModule._callAction (/Users/amar/dev/personal/crypto/nft-rental-marketplace/node_modules/hardhat/src/internal/hardhat-network/provider/modules/eth.ts:353:9)
      at HardhatNetworkProvider._sendWithLogging (/Users/amar/dev/personal/crypto/nft-rental-marketplace/node_modules/hardhat/src/internal/hardhat-network/provider/provider.ts:138:22)
      at HardhatNetworkProvider.request (/Users/amar/dev/personal/crypto/nft-rental-marketplace/node_modules/hardhat/src/internal/hardhat-network/provider/provider.ts:115:18)
      at JsonRpcHandler._handleRequest (/Users/amar/dev/personal/crypto/nft-rental-marketplace/node_modules/hardhat/src/internal/hardhat-network/jsonrpc/handler.ts:188:20)
      at JsonRpcHandler._handleSingleRequest (/Users/amar/dev/personal/crypto/nft-rental-marketplace/node_modules/hardhat/src/internal/hardhat-network/jsonrpc/handler.ts:167:17)
      at Server.JsonRpcHandler.handleHttp (/Users/amar/dev/personal/crypto/nft-rental-marketplace/node_modules/hardhat/src/internal/hardhat-network/jsonrpc/handler.ts:52:21)

eth_sendRawTransaction
  Contract call:       <UnrecognizedContract>
  Transaction:         0x590067620dfd03c85037ebd1fcb890e6a90c165828c3a6d5f7713f90167e78e0
  From:                0x976ea74026e726554db657fa54763abd0c3a0aa9
  To:                  0xf953b3a269d80e3eb0f2947630da976b896a8c5b
  Value:               0 ETH
  Gas used:            29345 of 29345
  Block #176:          0x9f90ce13745b5dc573a768ebe23c5faf5a3ee518f0239dd3feec223fc894429f
```
Right as metamask pops up for the approval, and before anything is selected by the user.

Comment (forcey) [reply src/routes/return.js:13]: I tried your branch and there was no problem approving

image

Try to deploy again? I think your local contract address might be incorrect.

Comment (forcey) [reply src/routes/return.js:13]: Note that you should approve Kasu contract's the address, not the lender's address.

**Comment by forcey** [comment]
Amar, I'm going to merge this, since I'm building on top of your ERC721 abis.
I've also implemented a reusable approval checker, and I'll integrate it with your dialog. Feel free to add additional improvements as separate PRs.

**Pull request merged**